dname.c

```diff
@@ -620,3 +620,107 @@ is_dname_subdomain_of_case(const uint8_t* d, unsigned int len,
 	/* The trailing portion is not at a label point. */
 	return 0;
 }
+
+size_t
+buf_dname_length(const uint8_t* buf, size_t len)
+{
+	size_t l = 0;
+	if(!buf || len == 0)
+		return l;
+	while(len > 0 && buf[0] != 0) {
+		size_t lablen = (size_t)(buf[0]);
+		if( (lablen&0xc0) )
+			return 0; /* the name should be uncompressed */
+		if(lablen+1 > len)
+			return 0; /* should fit in the buffer */
+		l += lablen+1;
+		len -= lablen+1;
+		buf += lablen+1;
+		if(l > MAXDOMAINLEN)
+			return 0;
+	}
+	if(len == 0)
+		return 0; /* end label should fit in buffer */
+	if(buf[0] != 0)
+		return 0; /* must end in root label */
+	l += 1; /* for the end root label */
+	if(l > MAXDOMAINLEN)
+		return 0;
+	return l;
+}
+
+int
+dname_make_buffered(struct dname_buffer* dname, uint8_t *name, int normalize)
+{
+	size_t name_size = 0;
+	uint8_t label_offsets[MAXDOMAINLEN];
```

Comment:
Could this be `MAXDOMAINLEN/2+1`?

```suggestion
	uint8_t label_offsets[MAXDOMAINLEN/2+1];
```

Reply:
Yes it looks like it should be. It was copied from dname_make() earlier, that had the original max.

Reply:
In https://github.com/NLnetLabs/nsd/pull/444/commits/48fde5528f09409ee3228264840a836e6eaff6ee it also fixes dname_make in the same way.